libopeniscsiusr/idbm.c

```diff
@@ -0,0 +1,699 @@
+}
+
+static int _idbm_recs_read(struct iscsi_context *ctx, struct idbm_rec *recs,
+			   const char *conf_path)
+{
+	int rc = LIBISCSI_OK;
+	char name[NAME_MAXVAL];
+	char value[VALUE_MAXVAL];
+	char *line = NULL;
+	char *nl = NULL;
+	char buffer[2048];
+	int line_number = 0;
+	int c = 0;
+	int i = 0;
+	FILE *f = NULL;
+	int errno_save = 0;
+	char strerr_buff[_STRERR_BUFF_LEN];
+
+	assert(ctx != NULL);
+	assert(recs != NULL);
+	assert(conf_path != NULL);
+
+	f = fopen(conf_path, "r");
+	errno_save = errno;
+	if (!f) {
+		_error(ctx, "Failed to open %s using read mode: %d %s",
+		       conf_path, errno_save,
+		       _strerror(errno_save, strerr_buff));
+		rc = LIBISCSI_ERR_IDBM;
+		goto out;
+	}
+
+	_info(ctx, "Parsing iSCSI interface configuration %s", conf_path);
+	/* process the config file */
+	do {
+		line = fgets(buffer, sizeof (buffer), f);
+		line_number++;
+		if (!line)
+			continue;
+		if (strlen(line) == 0)
+			continue;
+
+		nl = line + strlen(line) - 1;
+		if (*nl != '\n') {
+			_warn(ctx, "Config file %s line %d too long.",
+			      conf_path, line_number);
+			continue;
+		}
+
+		line = strstrip(line);
+		/* process any non-empty, non-comment lines */
+		if (!*line || *line == '\0' || *line ==  '\n' || *line == '#')
+			continue;
+
+		/* parse name */
+		i=0; nl = line; *name = 0;
+		while (*nl && !isspace(c = *nl) && *nl != '=') {
+			*(name+i) = *nl; i++; nl++;
+		}
+		if (!*nl) {
+			_warn(ctx, "config file %s line %d do not has value",
+			      conf_path, line_number);
+			continue;
+		}
+		*(name+i)=0; nl++;
+		/* skip after-name traling spaces */
+		while (*nl && isspace(c = *nl)) nl++;
+		if (*nl && *nl != '=') {
+			_warn(ctx, "config file %s line %d has not '=' "
+			      "separator", conf_path, line_number);
+			continue;
+		}
+		/* skip '=' sepa */
+		nl++;
+		/* skip after-sepa traling spaces */
+		while (*nl && isspace(c = *nl)) nl++;
+		if (!*nl) {
+			_warn(ctx, "config file %s line %d do not has value",
+			      conf_path, line_number);
+			continue;
+		}
+		/* parse value */
+		i=0; *value = 0;
+		while (*nl) {
+			*(value+i) = *nl; i++; nl++;
+		}
+		*(value+i) = 0;
+
+		rc = _idbm_rec_update_param(ctx, recs, name, value,
+					    line_number);
+		if (rc == LIBISCSI_ERR_INVAL) {
+			_error(ctx, "config file %s invalid.", conf_path);
+			goto out;
+		} else if (rc != LIBISCSI_OK)
+		      goto out;
+	} while (line);
+
+out:
+	if (f != NULL)
+		fclose(f);
+	return rc;
+}
+
+int _idbm_iface_get(struct iscsi_context *ctx, const char *iface_name, struct
+		    iscsi_iface **iface)
+{
+	int rc = LIBISCSI_OK;
+	char conf_path[PATH_MAX];
+	struct idbm_rec *recs = NULL;
+
+	assert(iface != NULL);
+	assert(ctx != NULL);
+
+	*iface = NULL;
+
+	if (iface_name == NULL)
+		goto out;
+
+	snprintf(conf_path, PATH_MAX, "%s/%s", IFACE_CONFIG_DIR, iface_name);
+
+	*iface = calloc(1, sizeof(struct iscsi_iface));
+	_alloc_null_check(ctx, *iface, rc, out);
+
+	snprintf((*iface)->name, sizeof((*iface)->name)/sizeof(char),
```

Comment:
Would it be clearer to use constant ISCSI_MAX_IFACE_LEN instead?

Reply:
All of the `snprintf` calls are not checking the return value.  I think it would be good to at least ensure they are all null terminated.

Reply:
Using the constant might cause human error(incorrectly chose array length), runtime size calculation does not.
According to man 3 of snprintf, it will always be null terminated:
`The functions snprintf() and vsnprintf() write at most size bytes (including the terminating null byte ('\0')) to str.`

Reply:
Oops, you are correct, I was thinking of `strncpy`